wayland-backend/tests/rs_sys_impls.rs

```diff
@@ -0,0 +1,167 @@
+///
+/// ```
+/// #[test]
+/// fn test() {
+///     assert_impl!(server::ObjectId: std::fmt::Debug);
+/// }
+/// ```
+///
+/// Multiple traits may be tested by separating each trait with a comma.
+///
+/// ```
+/// #[test]
+/// fn test() {
+///     assert_impl!(server::ObjectId: std::fmt::Debug, Send, Sync);
+/// }
+/// ```
+///
+/// For the client side, simply change the path before the name of the type.
+///
+/// ```
+/// #[test]
+/// fn test() {
+///     assert_impl!(client::ObjectId: std::fmt::Debug)
+/// }
+/// ```
+///
+/// Traits may be tested through prefixing the contents of the macro with `dyn`.
+///
+/// ```ignore
+/// #[test]
+/// fn test() {
+///     assert_impl!(dyn server::SomeTraitWithNoGeneric: std::fmt::Debug)
+/// }
+/// ```
+///
+/// Finally, generics may also be tested by simply adding the generics as expected in a normal type. Do note
+/// you will need to monomorphize the type with something, such as, `()`, the unit type.
+///
+/// ```
+/// #[test]
+/// fn test() {
+///     assert_impl!(server::Backend<()>: Send, Sync); // No trait
+///     assert_impl!(dyn server::ObjectData<()>: std::fmt::Debug); // Trait
+/// }
+/// ```
+macro_rules! assert_impl {
+    // No type parameters with dyn
+    (
+        dyn $side: ident::$ty: ident: $($trait_: path),+
+    ) => {{
+        #[allow(dead_code)]
+        fn assert_impl<T: $($trait_ +)* ?Sized>() {}
+
+        assert_impl::<dyn wayland_backend::rs::$side::$ty>();
+        #[cfg(feature = "server_system")]
+        assert_impl::<dyn wayland_backend::sys::$side::$ty>();
+    }};
+
+    // Type parameters with dyn
+    (
+        dyn $side: ident::$ty: ident<$($types: ty),*>: $($trait_: path),+
+    ) => {{
+        #[allow(dead_code)]
+        fn assert_impl<T: $($trait_ +)* ?Sized>() {}
+
+        assert_impl::<dyn wayland_backend::rs::$side::$ty <$($types),*>>();
+        #[cfg(feature = "server_system")]
+        assert_impl::<dyn wayland_backend::sys::$side::$ty <$($types),*>>();
+    }};
+
+    // No type parameters
+    (
+        $side: ident::$ty: ident: $($trait_: path),+
+    ) => {{
+        #[allow(dead_code)]
+        fn assert_impl<T: $($trait_ +)* ?Sized>() {}
+
+        assert_impl::<wayland_backend::rs::$side::$ty>();
+        #[cfg(feature = "server_system")]
+        assert_impl::<wayland_backend::sys::$side::$ty>();
+    }};
+
+    // Type parameters
+    (
+        $side: ident::$ty: ident<$($types: ty),*>: $($trait_: path),+
+    ) => {{
+        #[allow(dead_code)]
+        fn assert_impl<T: $($trait_ +)* ?Sized>() {}
+
+        assert_impl::<wayland_backend::rs::$side::$ty <$($types),*>>();
+        #[cfg(feature = "server_system")]
+        assert_impl::<wayland_backend::sys::$side::$ty <$($types),*>>();
+    }};
+}
+
+mod server {
+    #[test]
+    fn test_impls() {
+        // ObjectData
+        assert_impl!(
+            dyn server::ObjectData<()>: downcast_rs::DowncastSync
+        );
+        #[cfg(not(tarpaulin_include))]
```

Comment:
Why are those annotations here?

Reply:
Does tarpaulin not apply in testing land?

Reply:
gone